File: area_scan_3d_camera/Basic/CapturePointCloudWithNormals/CapturePointCloudWithNormals.cpp
/*******************************************************************************
 *BSD 3-Clause License
 *
 *Copyright (c) 2016-2024, Mech-Mind Robotics
 *All rights reserved.
 *
 *Redistribution and use in source and binary forms, with or without
 *modification, are permitted provided that the following conditions are met:
 *
 *1. Redistributions of source code must retain the above copyright notice, this
 *   list of conditions and the following disclaimer.
 *
 *2. Redistributions in binary form must reproduce the above copyright notice,
 *   this list of conditions and the following disclaimer in the documentation
 *   and/or other materials provided with the distribution.
 *
 *3. Neither the name of the copyright holder nor the names of its
 *   contributors may be used to endorse or promote products derived from
 *   this software without specific prior written permission.
 *
 *THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 *AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 *IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 *DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 *FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 *DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 *SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 *CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 *OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 *OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 ******************************************************************************/

/*
With this sample, you can calculate normals and save the point cloud with normals. The normals can be calculated on the camera or the computer.
*/

#include <iostream>
#include "area_scan_3d_camera/Camera.h"
#include "area_scan_3d_camera/api_util.h"

int main()
{
    mmind::eye::Camera camera;
    if (!findAndConnect(camera))
        return -1;

    if (!confirmCapture3D()) {
        camera.disconnect();
        return 0;
    }

    // Calculate the normals of the points on the camera and save the point cloud with normals to file
    mmind::eye::Frame3D frame3D;
    if (camera.capture3DWithNormal(frame3D).isOK()) {
        showError(frame3D.saveUntexturedPointCloudWithNormals(mmind::eye::FileFormat::PLY,
                                                              "PointCloud_1.ply"));
    } else {
        std::cout << "Failed to capture the point cloud." << std::endl;
        camera.disconnect();
        return -1;
    }

    // Calculate the normals of the points on the computer and save the point cloud with normals to file
    if (camera.capture3D(frame3D).isOK()) {
        showError(frame3D.saveUntexturedPointCloudWithNormals(mmind::eye::FileFormat::PLY,
                                                              "PointCloud_2.ply"));
    } else {
        std::cout << "Failed to capture the point cloud." << std::endl;
        camera.disconnect();
        return -1;
    }

    camera.disconnect();
    std::cout << "Disconnected from the camera successfully." << std::endl;

    return 0;
}

File: area_scan_3d_camera/Basic/capture_point_cloud_with_normals.py
# With this sample, you can calculate normals and save the point cloud with normals. The normals can be calculated on the camera or the computer.

from mecheye.shared import *
from mecheye.area_scan_3d_camera import *
from mecheye.area_scan_3d_camera_utils import find_and_connect, confirm_capture_3d


class CapturePointCloudWithNormals(object):
    def __init__(self):
        self.camera = Camera()
        self.frame_3d = Frame3D()

    # Calculate the normals of the points on the camera and save the point cloud with normals to file
    def capture_point_cloud_with_normals_calculated_on_camera(self):
        point_cloud_file = "PointCloud_1.ply"
        if self.camera.capture_3d_with_normal(self.frame_3d).is_ok():
            show_error(
                self.frame_3d.save_untextured_point_cloud_with_normals(FileFormat_PLY, point_cloud_file))
            return True
        else:
            print("Failed to capture the point cloud.")
            self.camera.disconnect()
            print("Disconnected from the camera successfully.")
            return False

    # Calculate the normals of the points on the computer and save the point cloud with normals to file
    def capture_point_cloud_with_normals_calculated_locally(self):
        point_cloud_file = "PointCloud_2.ply"
        if self.camera.capture_3d(self.frame_3d).is_ok():
            show_error(
                self.frame_3d.save_untextured_point_cloud_with_normals(FileFormat_PLY, point_cloud_file))
            return True
        else:
            print("Failed to capture the point cloud.")
            self.camera.disconnect()
            print("Disconnected from the camera successfully.")
            return False

    def main(self):
        if find_and_connect(self.camera):
            if not confirm_capture_3d():
                return
            if not self.capture_point_cloud_with_normals_calculated_on_camera():
                return
            if not self.capture_point_cloud_with_normals_calculated_locally():
                return
            self.camera.disconnect()
            print("Disconnected from the camera successfully.")


if __name__ == '__main__':
    a = CapturePointCloudWithNormals()
    a.main()
